src/bun.js/ipc.zig

```diff
@@ -156,17 +156,162 @@ pub const IPCData = struct {
+        }
+    }
+
+    fn uvWriteCallback(req: *uv.uv_write_t, status: uv.ReturnCode) callconv(.C) void {
+        const this = bun.cast(*NamedPipeIPCData, req.data);
+        log("uvWriteCallback {d} {d} {d}", .{ status.int(), this.current_payload_len, this.outgoing.list.len });
+        defer bun.destroy(req);
+        if (status.errEnum()) |_| {
+            Output.printErrorln("Failed write IPC data", .{});
+            return;
+        }
+        const n = this.current_payload_len;
+        if (n == this.outgoing.list.len) {
+            this.outgoing.cursor = 0;
+            this.outgoing.list.len = 0;
+        } else {
+            this.outgoing.cursor += @intCast(n);
+            this.processSend();
+        }
+    }
+
+    pub fn writeVersionPacket(this: *NamedPipeIPCData) void {
+        if (Environment.allow_assert) {
+            std.debug.assert(this.has_written_version == 0);
+        }
+        const VersionPacket = extern struct {
+            type: IPCMessageType align(1) = .Version,
+            version: u32 align(1) = ipcVersion,
+        };
+
+        if (Environment.allow_assert) {
+            this.has_written_version = 1;
+        }
+        const bytes = comptime std.mem.asBytes(&VersionPacket{});
+        // enqueue to be sent after connecting
+        var list = this.outgoing.list.listManaged(bun.default_allocator);
+        list.appendSlice(bytes) catch bun.outOfMemory();
+        if (this.connected) {
+            this.processSend();
+        }
+    }
+
+    pub fn serializeAndSend(this: *NamedPipeIPCData, globalThis: *JSGlobalObject, value: JSValue) bool {
+        if (Environment.allow_assert) {
+            std.debug.assert(this.has_written_version == 1);
+        }
+
+        const serialized = value.serialize(globalThis) orelse return false;
+        defer serialized.deinit();
+
+        const size: u32 = @intCast(serialized.data.len);
+        log("serializeAndSend {d}", .{size});
+
+        const payload_length: usize = @sizeOf(IPCMessageType) + @sizeOf(u32) + size;
+
+        this.outgoing.list.ensureUnusedCapacity(bun.default_allocator, payload_length) catch @panic("OOM");
+        const start_offset = this.outgoing.list.len;
+
+        this.outgoing.list.writeTypeAsBytesAssumeCapacity(u8, @intFromEnum(IPCMessageType.SerializedMessage));
+        this.outgoing.list.writeTypeAsBytesAssumeCapacity(u32, size);
+        this.outgoing.list.appendSliceAssumeCapacity(serialized.data);
+
+        std.debug.assert(this.outgoing.list.len == start_offset + payload_length);
+
+        if (start_offset == 0) {
+            std.debug.assert(this.outgoing.cursor == 0);
+            if (this.connected) {
+                this.processSend();
+            }
+        }
+
+        return true;
+    }
+
+    pub fn close(this: *NamedPipeIPCData, comptime Context: type) void {
+        if (this.server.loop != null) {
+            _ = uv.uv_close(@ptrCast(&this.pipe), NewNamedPipeIPCHandler(Context).onServerClose);
```

Comment:
This should be a function on the `this.pipe`, not calling libuv internals directly.